### PR TITLE
Draft support for cancelling test execution

### LINKED ISSUE #1880: Introduce `pleaseStop()` method for `org.junit.platform.launcher.Launcher`
We would like to send a command to the engine to control the execution and stop the progress of pending tests via the interface `org.junit.platform.launcher.Launcher`.

### PR DESCRIPTION
**Pull request by marcphilipp**
## Overview

* Introduce `CancellationToken` interface and default implementation
* Add `Launcher.execute(TestPlan, CancellationToken, ...)` method
* Add `ExecutionRequest.getCancellationToken()` method
* Implement support for cancellation for engines extending
  `HierarchicalTestEngine`

Eventually resolves #1880.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://docs.junit.org/snapshot/user-guide/) and [Release Notes](https://docs.junit.org/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/Launcher.java [hunk 1]
```diff
@@ -127,4 +129,25 @@ public interface Launcher {
...
+	 *
+	 * @apiNote The supplied {@link TestPlan} must not have been executed
+	 * previously.
+	 *
+	 * @param testPlan the test plan to execute; never {@code null}
+	 * @param listeners additional test execution listeners; never {@code null}
+	 * @since 6.0
+	 */
+	@API(status = EXPERIMENTAL, since = "6.0")
+	void execute(TestPlan testPlan, CancellationToken cancellationToken, TestExecutionListener... listeners);
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/Launcher.java:185]: New API for build tools and IDEs

File: junit-platform-engine/src/main/java/org/junit/platform/engine/ExecutionRequest.java [hunk 2]
```diff
@@ -171,4 +174,8 @@ public NamespacedHierarchicalStore<Namespace> getStore() {
 			"No NamespacedHierarchicalStore was configured for this request");
 	}
 
+	public CancellationToken getCancellationToken() {
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/ExecutionRequest.java:177]: New API for `TestEngine` implementors

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTestTask.java [hunk 3]
```diff
@@ -139,8 +142,10 @@ private void prepare() {
 		parentContext = null;
 	}
 
-	private void checkWhetherSkipped() {
-		requiredThrowableCollector().execute(() -> skipResult = node.shouldBeSkipped(requiredContext()));
+	private SkipResult checkWhetherSkipped() throws Exception {
+		return taskContext.cancellationToken().isCancellationRequested() //
+				? CANCELLED_SKIP_RESULT //
+				: node.shouldBeSkipped(requiredContext());
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTestTask.java:148]: Support for engines extending `HierarchicalTestEngine`

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineExecutionOrchestrator.java [hunk 4]
```diff
@@ -223,12 +228,18 @@ private ListenerRegistry<TestExecutionListener> buildListenerRegistryForExecutio
 	private void executeEngine(TestDescriptor engineDescriptor, EngineExecutionListener listener,
 			ConfigurationParameters configurationParameters, TestEngine testEngine,
 			OutputDirectoryProvider outputDirectoryProvider, DiscoveryIssueNotifier discoveryIssueNotifier,
-			NamespacedHierarchicalStore<Namespace> requestLevelStore) {
+			NamespacedHierarchicalStore<Namespace> requestLevelStore, CancellationToken cancellationToken) {
+
+		if (cancellationToken.isCancellationRequested()) {
+			listener.executionSkipped(engineDescriptor, "Cancellation of execution requested");
+			return;
+		}
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/EngineExecutionOrchestrator.java]: Skip entire engine if cancellation has already been requested. This avoid the overhead of initializing an engine's execution only to skip its root node. Moreover, this is at least a stop-gap solution for engines not supporting cancellation at all.

File: jupiter-tests/src/test/java/org/junit/jupiter/engine/ExecutionCancellationTests.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+		results.allEvents().assertEventsMatchExactly( //
+			event(engine(), started()), //
+			event(container(testClass), started()), //
+			event(test("first"), started()), //
+			event(test("first"), reportEntry(Map.of("cancelled", "true"))), //
+			event(test("first"), finishedSuccessfully()), //
+			event(test("second"), skippedWithReason("Test execution cancelled")), //
+			event(container(testClass), finishedSuccessfully()), //
+			event(engine(), finishedSuccessfully()));
```
Comment (marcphilipp) [inline jupiter-tests/src/test/java/org/junit/jupiter/engine/ExecutionCancellationTests.java:58]: This test demonstrates what happens if execution is cancelled during execution of the first test. The second test is reported as skipped and all later test classes would be as well.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/ExecutionRequest.java [hunk 6]
```diff
@@ -171,4 +174,8 @@ public NamespacedHierarchicalStore<Namespace> getStore() {
...
+		return Preconditions.notNull(this.cancellationToken, "No CancellationToken was configured for this request");
```
Comment (mpkorstanje) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/ExecutionRequest.java:178]: I had some difficulty inferring from what code how I could safely use the cancellation token. Instead of making the `cancellationToken` nullable, I would expect a default implementation to be provided in the deprecated  constructor.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTestTask.java (hunk 3 above)
Comment (mpkorstanje) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTestTask.java:148]: For Cucumber this means that cancellation will work out of the box. :+1:

File: junit-platform-suite-engine/src/main/java/org/junit/platform/suite/engine/SuiteTestEngine.java [hunk 7]
```diff
@@ -73,7 +73,7 @@ public void execute(ExecutionRequest request) {
 		suiteEngineDescriptor.getChildren()
 				.stream()
 				.map(SuiteTestDescriptor.class::cast)
-				.forEach(suiteTestDescriptor -> suiteTestDescriptor.execute(engineExecutionListener, requestLevelStore));
+				.forEach(suiteTestDescriptor -> suiteTestDescriptor.execute(engineExecutionListener, requestLevelStore, request.getCancellationToken()));
```
Comment (mpkorstanje) [inline junit-platform-suite-engine/src/main/java/org/junit/platform/suite/engine/SuiteTestEngine.java:76]: This will fail if a request was created with the deprecated constructor.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/ExecutionRequest.java (hunk 6 above)
Comment (marcphilipp) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/ExecutionRequest.java:178]: I think we can instead return a default one here since it has a sensible default.

File: junit-platform-suite-engine/src/main/java/org/junit/platform/suite/engine/SuiteTestEngine.java (hunk 7 above)
Comment (marcphilipp) [reply junit-platform-suite-engine/src/main/java/org/junit/platform/suite/engine/SuiteTestEngine.java:76]: `ExecutionRequest` really should be created by anyone outside of the Platform.

**Comment by marcphilipp** [comment]
Superseded by #4728 and #4725.
